syntax = "proto3";

//! Dummy message
message Empty {}

//! Stores all the information for an item
message ItemData
{
	//! The inventory ID of the item
	uint32 id = 1;

	//! The display name
	optional string name = 2;

	//! The unit cost
	optional float price = 3;

	//! Quantity of the item in stock
	optional float quantity = 4;
}

//! Response after trying to create new item
message NewItemResponse
{
	//! The ID of the new item allocated by the server
	uint32 new_id = 1;
}

//! Asks the server to delete an item
message DeleteItemRequest
{
	//! The ID of the item to be deleted
	uint32 delete_id = 1;
}

//! Describes the services provided by the inventory server
service InventoryService
{
	//! Adds a new item to the inventory
	rpc CreateItem(Empty) returns (NewItemResponse) {}

	//! Deletes an item from the inventory
	rpc DeleteItem(DeleteItemRequest) returns (Empty) {}

	//! Modifies the data of the item
	rpc ModifyItem(ItemData) returns (Empty) {}

	//! Reads a list of all items
	rpc Subscribe(Empty) returns (stream ItemData) {}
}

// The full database
message DataBase
{
	repeated ItemData Items = 1;
}